### PR TITLE
[aes] Only unmask final cipher core output

### PR DESCRIPTION
**Pull request by vogelpi**
This PR inserts a mux between the masked cipher core output and the data output register/feedback path to: 
1. Avoid SCA leakage occuring from unmasking intermediate cipher core outputs.
2. Increase the difficulty for extracting intermediate data via FI.

Adding this additional mux has been on my list for a very long time now but I couldn't do it as the SCA setup wasn't working. I have an SCA experiment running to verify that this is working as intended, but I think it can still be reviewed in the meantime.

### CODE REVIEW

File: hw/ip/aes/rtl/aes_core.sv
Comment (tjaychen) [inline hw/ip/aes/rtl/aes_core.sv:433]: @vogelpi this looks good to me, although i guess it's worth asking..do we want the final output to be unmasked in general?

I can't remember which security consultancy brought it up, but their point is that if we use software to wrap / unwrap keys, it's probably best to keep the output in shares also.

@moidx Adding Miguel to see if this is something we need to talk through a bit more, because there would definitely be a software impact on normal data usage.

Comment (vogelpi) [reply hw/ip/aes/rtl/aes_core.sv:433]: Thanks for the feedback @tjaychen . My understanding is that we won't provide the final output in shared/masked form as wrap/unwrap flows for AES are no longer on the table.

Comment (tjaychen) [reply hw/ip/aes/rtl/aes_core.sv:433]: wait really? let me double check.
I would have thought key wrap / unwrap would be a heavy use case in these systems.

**Review by vrozic — APPROVED** [review]
Looks good to me

**Comment by vogelpi** [comment]
Thanks for getting back @tjaychen . Yes, this is exactly what I meant - software can only do 128b blocks at a time. Glad to know we are in agreement.

File: hw/ip/aes/rtl/aes_core.sv
Comment (cdgori) [inline hw/ip/aes/rtl/aes_core.sv:444]: nit: I didn't look to see if you did this throughout for the manual buffer instantiations but I could make an argument for Width value being 128*NumShares (rather than 256)

**Review by cdgori — APPROVED** [review]
> Update: I've finished the FPGA experiment and couldn't identify a change in leakage. I think we should merge this. @cdgori would you mind having a look at this, too?

The change looks correct to me on paper - thanks for running the experiment since that's really the definitive thing.

Sorry this took me so long to get to.

File: hw/ip/aes/rtl/aes_core.sv
Comment (vogelpi) [reply hw/ip/aes/rtl/aes_core.sv]: Done.

**Comment by vogelpi** [comment]
Thanks @cdgori for reviewing and approving this PR.

Now all that's left is the approval from a committer. @tjaychen , since you already had a look, would mind approving? Regarding the unwrap/wrap case, I can open a separate issue or we can wait for more feedback on this. Just let me know what you prefer. Thanks!

**Comment by tjaychen** [comment]
sorry i didn't realize i hadn't pushed approve.

**Review by tjaychen — APPROVED** [review]

**Comment by vogelpi** [comment]
Thanks Tim.

**Pull request merged**
